File: mobile-native-backup/ios/AppSupport/InMemoryEncryptedStore.swift
import Foundation

struct StoredCipherMessage {
    let id: Int64
    let conversationID: Int64
    let senderID: Int64
    let receiverID: Int64
    let ciphertextB64: String
    let headerJSON: String
    let createdAt: Date
}

final class InMemoryEncryptedStore: EncryptedStore {
    private var messages: [StoredCipherMessage] = []
    private var syncDate: Date?
    private let lock = NSLock()

    func saveCiphertextMessage(id: Int64, conversationID: Int64, senderID: Int64, receiverID: Int64, ciphertextB64: String, headerJSON: String, createdAt: Date) throws {
        lock.lock()
        defer { lock.unlock() }
        guard messages.contains(where: { $0.id == id }) == false else {
            return
        }
        messages.append(
            StoredCipherMessage(
                id: id,
                conversationID: conversationID,
                senderID: senderID,
                receiverID: receiverID,
                ciphertextB64: ciphertextB64,
                headerJSON: headerJSON,
                createdAt: createdAt
            )
        )
    }

    func lastSyncTimestamp() throws -> Date? {
        lock.lock()
        defer { lock.unlock() }
        return syncDate
    }

    func setLastSyncTimestamp(_ date: Date) throws {
        lock.lock()
        defer { lock.unlock() }
        syncDate = date
    }

    func allMessages() -> [StoredCipherMessage] {
        lock.lock()
        defer { lock.unlock() }
        return messages
    }
}

File: mobile-native-backup/android/app/src/main/java/com/securemsg/storage/InMemoryEncryptedStore.kt
package com.securemsg.storage

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

data class StoredCipherMessage(
    val id: Long,
    val conversationId: Long,
    val senderId: Long,
    val receiverId: Long,
    val ciphertextB64: String,
    val headerJson: String,
    val createdAt: String
)

class InMemoryEncryptedStore : EncryptedStore {
    private val mutex = Mutex()
    private val messages = mutableListOf<StoredCipherMessage>()
    private var lastSync: String? = null

    override suspend fun saveCiphertextMessage(
        id: Long,
        conversationId: Long,
        senderId: Long,
        receiverId: Long,
        ciphertextB64: String,
        headerJson: String,
        createdAt: String
    ) {
        mutex.withLock {
            if (messages.any { it.id == id }) {
                return
            }
            messages += StoredCipherMessage(
                id = id,
                conversationId = conversationId,
                senderId = senderId,
                receiverId = receiverId,
                ciphertextB64 = ciphertextB64,
                headerJson = headerJson,
                createdAt = createdAt
            )
        }
    }

    override suspend fun getLastSyncTimestamp(): String? {
        return mutex.withLock { lastSync }
    }

    override suspend fun setLastSyncTimestamp(isoTimestamp: String) {
        mutex.withLock {
            lastSync = isoTimestamp
        }
    }

    suspend fun allMessages(): List<StoredCipherMessage> {
        return mutex.withLock { messages.toList() }
    }
}
